go/vt/vtctl/grpcvtctldserver/query.go

```diff
@@ -0,0 +1,300 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package grpcvtctldserver
+
+import (
+	"fmt"
+	"time"
+
+	"vitess.io/vitess/go/mysql/collations"
+	"vitess.io/vitess/go/protoutil"
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/topo/topoproto"
+	"vitess.io/vitess/go/vt/vtctl/schematools"
+
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	vtctldatapb "vitess.io/vitess/go/vt/proto/vtctldata"
+	"vitess.io/vitess/go/vt/proto/vttime"
+)
+
+// 2023-08-05 07:15:16
+const mysqlTimestampLayout = "2006-01-02 15:04:05"
+
+// TODO: explicit column list to guarantee parsing stability
+func selectSchemaMigrationsQuery(condition, order, skipLimit string) string {
+	return fmt.Sprintf(`select
+		*
+		from _vt.schema_migrations where %s %s %s`, condition, order, skipLimit)
+}
+
+func rowToSchemaMigration(row sqltypes.Row) (sm *vtctldatapb.SchemaMigration, err error) {
+	row = row[1:] // We're off by 1 because of the extra Tablet field added to the grouped query result
+
+	sm = new(vtctldatapb.SchemaMigration)
+	sm.Uuid = row[1].ToString()
+	sm.Keyspace = row[2].ToString()
+	sm.Shard = row[3].ToString()
+	sm.Schema = row[4].ToString()
+	sm.Table = row[5].ToString()
+	sm.MigrationStatement = row[6].ToString()
+
+	if name := row[7].ToString(); name != "" {
+		sm.Strategy, err = schematools.ParseSchemaMigrationStrategy(name)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	sm.Options = row[8].ToString()
+
+	sm.AddedAt, err = valueToVTTime(row[9])
+	if err != nil {
+		return nil, err
+	}
+
+	sm.RequestedAt, err = valueToVTTime(row[10])
+	if err != nil {
+		return nil, err
+	}
+
+	sm.ReadyAt, err = valueToVTTime(row[11])
+	if err != nil {
+		return nil, err
+	}
+
+	sm.StartedAt, err = valueToVTTime(row[12])
+	if err != nil {
+		return nil, err
+	}
+
+	// liveness_indicator row[13]
+
+	sm.CompletedAt, err = valueToVTTime(row[14])
+	if err != nil {
+		return nil, err
+	}
+
+	sm.CleanedUpAt, err = valueToVTTime(row[15])
+	if err != nil {
+		return nil, err
+	}
+
+	if status := row[16].ToString(); status != "" {
+		sm.Status, err = schematools.ParseSchemaMigrationStatus(status)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	sm.LogPath = row[17].ToString()
+	sm.Artifacts = row[18].ToString()
+
+	sm.Retries, err = row[19].ToUint64()
+	if err != nil {
+		return nil, err
+	}
+
+	if alias := row[20].ToString(); alias != "" {
+		sm.Tablet, err = topoproto.ParseTabletAlias(alias)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	sm.TabletFailure, err = row[21].ToBool()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.Progress, err = row[22].ToFloat32()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.MigrationContext = row[23].ToString()
+	// ddl_action row[24]
```

Comment:
also still missing because i'm debating whether it's worth defining an enum for or just dealing with the raw string

Reply:
IMHO raw string is fine.